### PR TITLE
INGM-407: make new example about can bootloader

### PR DESCRIPTION
**Pull request by WoodElv**
### Description

A new script is added into the examples folder for loading a firmware using CAN bootloader.
Also, unit tests are added for this script.

### Type of change

Please add a description and delete options that are not relevant.

- [x] Script for loading a firmware using CAN bootloader.

### Tests
- [x] Unit test if the script successes.
- [x] Unit test if the script failed.

Please describe the tests that you ran to verify your changes if it applies. 

### Documentation

Please update the documentation.

- [x] USe [type hints](https://docs.python.org/3/library/typing.html) for every function and argument.
- [x] Build documentation locally to verify changes.

### Code formatting

- [x] Use black package to format the code: `black -l 100 ingeniamotion tests`. It is recommended to configure the code editor to automatically format the code using black with a max length line of 100.

### CODE REVIEW

File: examples/load_fw_canopen.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+def status_log(current_status: str) -> None:
+    print(f"Load firmware status: {current_status}")
+
+
+def progress_log(current_progress: int) -> None:
+    print(f"Load firmware progress: {current_progress}%")
+
+
+def load_firmware_canopen(can_drive: Dict[str, Any]) -> None:
+    # Create MotionController instance
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:18]: This comment does not provide much information, it could be removed.

File: examples/load_fw_canopen.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    mc = MotionController()
+    # Find available nodes
+    node_id_list = mc.communication.scan_servos_canopen(
+        can_drive["device"],
+        baudrate=can_drive["baudrate"],
+        channel=can_drive["channel"],
+    )
+    if can_drive["node_id"] not in node_id_list:
+        print("No nodes are connected.")
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:27]: This message is not correct. e.g.:
`print(f"Node {can_drive["node_id"]} is not detected.")`

File: examples/load_fw_canopen.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        return
+    else:
+        print(f"Found nodes: {node_id_list}")
+
+    # Connect drive
+    print("Starts to established a communication.")
+    alias_can_servo = "custom_alias"
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:34]: Why do you use a custom alias? Could not be the default one for simplicity?

File: examples/load_fw_canopen.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        channel=can_drive["channel"],
+        alias=alias_can_servo,
+    )
+    if alias_can_servo in mc.servos:
+        print("Drive is connected.")
+    else:
+        print("Drive is not connected.")
+        return
+
+    # Load firmware
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:49]: Remove comment.

File: examples/load_fw_canopen.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    try:
+        mc.communication.load_firmware_canopen(
+            can_drive["fw_path"],
+            servo=alias_can_servo,
+            status_callback=status_log,
+            progress_callback=progress_log,
+        )
+        print("Firmware is uploaded successfully.")
+    except ILFirmwareLoadError as e:
+        print(f"CAN boot error: {e}")
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:60]: Firmware loading failed.

File: examples/load_fw_canopen.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    finally:
+        try:
+            mc.communication.disconnect(alias_can_servo)
+            print("Drive is disconnected.")
+        except Exception as e:
+            print(f"Error when disconnection from drive: {e}")
```
Comment (pzinemanas) [inline examples/load_fw_canopen.py:66]: Error during drive disconnection?

File: examples/load_fw_canopen.py (hunk 3 above)
Comment (WoodElv) [reply examples/load_fw_canopen.py:34]: It's easier without a custom alias, but we don't have a lot of examples using a custom alias. I think it's good having some to teach how to use them. What do you think?

File: examples/load_fw_canopen.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            print(f"Error during drive disconnection: {e}")
+
+
+if __name__ == "__main__":
+    # Remember to replace all parameters here
+    can_drive = {
+        "device": CAN_DEVICE.KVASER,
+        "channel": 0,
+        "baudrate": CAN_BAUDRATE.Baudrate_1M,
+        "node_id": 20,
```
Comment (MartinALL91) [inline examples/load_fw_canopen.py:72]: I would set the node_id to the default value of our drives (32).

File: examples/load_fw_canopen.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        baudrate=can_drive["baudrate"],
+        channel=can_drive["channel"],
+    )
+    if can_drive["node_id"] not in node_id_list:
+        print(f"Node {can_drive['node_id']} is not detected.")
+        return
+    else:
+        print(f"Found nodes: {node_id_list}")
+
+    print("Starts to established a communication.")
```
Comment (MartinALL91) [inline examples/load_fw_canopen.py:31]: establish

**Review by pzinemanas — APPROVED** [review]

File: examples/load_fw_canopen.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    print("Starts to establish a communication.")
+    mc.communication.connect_servo_canopen(
+        can_drive["device"],
+        can_drive["dictionary_path"],
+        can_drive["node_id"],
+        baudrate=can_drive["baudrate"],
+        channel=can_drive["channel"],
+    )
+    if "default" in mc.servos:
```
Comment (mjedieckmann) [inline examples/load_fw_canopen.py:39]: should probably be try / catch instead

**Review by mjedieckmann — APPROVED** [review]

File: examples/load_fw_canopen.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    # Remember to replace all parameters here
+    can_drive = {
+        "device": CAN_DEVICE.KVASER,
+        "channel": 0,
+        "baudrate": CAN_BAUDRATE.Baudrate_1M,
+        "node_id": 32,
+        "dictionary_path": "parent_directory/full_dictionary_path.xdf",
+        "fw_path": "parent_directory/full_firmware_file_path.lfu",
+    }
+    load_firmware_canopen(can_drive)
```
Comment (GullinBustin) [inline examples/load_fw_canopen.py:73]: Why do you use a dict instead use different arguments?

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: examples/load_fw_canopen.py (hunk 10 above)
Comment (WoodElv) [reply examples/load_fw_canopen.py:73]: Because there are a lot of parameters, and I think it's cleaner passing just one input parameter instead of 6 parameters for a script.

Comment (GullinBustin) [reply examples/load_fw_canopen.py:73]: In my opinion, it is weird. A user can't know which parameters have the function seeing the declaration, an example of use is necessary to know all the parameters or see the full implementation, which is worse. A lot of customers will use these examples as templates to do their code, we must follow good practices.

File: tests/test_examples.py [hunk 11]
```diff
@@ -212,3 +214,89 @@ def enable_brake(*args, **kwargs):
     mocker.patch.object(MotionController, "configuration", MockConfiguration)
     result = script_runner.run(script_path, override, dictionary, f"-ip={ip_address}")
     assert result.returncode == 0
+
+
+@pytest.mark.virtual
+def test_can_bootloader_example_success(mocker, capsys):
+    device = "my_can_device"
+    channel = 111
+    baudrate = 4444
```
Comment (MartinALL91) [inline tests/test_examples.py:223]: These values are weird and in some cases make no sense. I would use normal ones.

File: examples/load_fw_canopen.py (hunk 9 above)
Comment (WoodElv) [reply examples/load_fw_canopen.py:40]: I'm going to remove this conditional statement because the no connection state can't be reached.

File: examples/load_fw_canopen.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            progress_callback=progress_log,
+        )
+        print("Firmware is uploaded successfully.")
+    except ILFirmwareLoadError as e:
+        print(f"Firmware loading failed: {e}")
+    finally:
+        try:
+            mc.communication.disconnect()
+            print("Drive is disconnected.")
+        except Exception as e:
```
Comment (GullinBustin) [inline examples/load_fw_canopen.py:56]: Why did you add a try-except in the disconnect function? and why is it a generic exception?

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: examples/load_fw_canopen.py (hunk 12 above)
Comment (WoodElv) [reply examples/load_fw_canopen.py:56]: I put it because if there was an unhandled exception. But I'm going to take it off because in that cases an exception should to be raised.

**Review by GullinBustin — APPROVED** [review]
